components/raftstore/src/store/fsm/apply.rs

```diff
@@ -3151,189 +2894,164 @@ where
-                Ok(msg) => self.msg_buf.push(msg),
-                Err(TryRecvError::Empty) => {
-                    expected_msg_count = Some(0);
-                    break;
-                }
-                Err(TryRecvError::Disconnected) => {
-                    normal.delegate.stopped = true;
-                    expected_msg_count = Some(0);
-                    break;
-                }
-            }
-        }
-        normal.handle_tasks(&mut self.apply_ctx, &mut self.msg_buf);
-        if normal.delegate.wait_merge_state.is_some() {
-            // Check it again immediately as catching up logs can be very fast.
-            expected_msg_count = Some(0);
-        } else if normal.delegate.yield_state.is_some() {
-            // Let it continue to run next time.
-            expected_msg_count = None;
-        }
-        expected_msg_count
-    }
-
-    fn end(&mut self, fsms: &mut [Box<ApplyFsm<EK>>]) {
-        let is_synced = self.apply_ctx.flush();
-        if is_synced {
-            for fsm in fsms {
-                fsm.delegate.last_sync_apply_index = fsm.delegate.apply_state.get_applied_index();
-            }
+    TLS_APPLY_CONTEXT.with(|e| unsafe {
+        if (*e.get()).is_null() {
+            return None;
         }
-    }
-}
-
-pub struct Builder<EK: KvEngine, W: WriteBatch<EK>> {
-    tag: String,
-    cfg: Arc<VersionTrack<Config>>,
-    coprocessor_host: CoprocessorHost<EK>,
-    importer: Arc<SSTImporter>,
-    region_scheduler: Scheduler<RegionTask<<EK as KvEngine>::Snapshot>>,
-    engine: EK,
-    sender: Box<dyn Notifier<EK>>,
-    router: ApplyRouter<EK>,
-    _phantom: PhantomData<W>,
-    store_id: u64,
-    pending_create_peers: Arc<Mutex<HashMap<u64, (u64, bool)>>>,
+        let ctx = &mut *(*e.get() as *mut Option<ApplyContext<EK, W>>);
+        ctx.take()
+    })
 }
 
-impl<EK: KvEngine, W> Builder<EK, W>
+pub fn flush_tls_ctx<EK, W>()
 where
+    EK: KvEngine,
     W: WriteBatch<EK>,
 {
-    pub fn new<T, C, ER: RaftEngine>(
-        builder: &RaftPollerBuilder<EK, ER, T, C>,
-        sender: Box<dyn Notifier<EK>>,
-        router: ApplyRouter<EK>,
-    ) -> Builder<EK, W> {
-        Builder {
-            tag: format!("[store {}]", builder.store.get_id()),
-            cfg: builder.cfg.clone(),
-            coprocessor_host: builder.coprocessor_host.clone(),
-            importer: builder.importer.clone(),
-            region_scheduler: builder.region_scheduler.clone(),
-            engine: builder.engines.kv.clone(),
-            _phantom: PhantomData,
-            sender,
-            router,
-            store_id: builder.store.get_id(),
-            pending_create_peers: builder.pending_create_peers.clone(),
+    TLS_APPLY_CONTEXT.with(|e| unsafe {
+        if (*e.get()).is_null() {
+            return;
         }
-    }
+        let ctx = &mut *(*e.get() as *mut Option<ApplyContext<EK, W>>);
+        if let Some(apply_ctx) = ctx.as_mut() {
+            if !apply_ctx.flush_notifier.is_empty() {
```

Comment:
Should we also drain the flush callback here?

Reply:
Why? `fn flush` will flush all callback.